examples/cfq/utils.py

```diff
@@ -0,0 +1,234 @@
+            for feature in input_features:
+                yield example[feature]
+
+
+def build_vocabulary(file_name: Text,
+                     input_features: Dict,
+                     tokenizer: text.Tokenizer,
+                     datasets: Iterable[tf.data.Dataset],
+                     special_tokens: Sequence[bytes] = (constants.PAD,
+                                                        constants.UNK,
+                                                        constants.BOS,
+                                                        constants.EOS),
+                     preprocessing_fun: Any = lambda x: x) -> Dict[bytes, int]:
+    """Returns a vocabulary of tokens with optional minimum frequency.
+    The vocabulary is saved to a temps file
+
+    Args:
+        
+        special_tokens: Special tokens that will be the start of the vocabulary.
+        tokenizer: tf tokenizer (eg. WhitespaceTokenizer)
+        datasets: datasets from which the vocabulary should be extracted
+        special_tokens: special tokens to be added in the vocabulary; they are
+                        added at the begginig with PAD at index 0
+        preprocessing_fun: function for applying preprocessing on a dataset
+                           entry; by default the identity function
+                           (no preprocessing)
+    Returns:
+        An ordered dictionary that maps tokens to their IDs in the vocabulary.
+    """
+
+    vocab_dir = 'temp'
+    vocab_path = vocab_dir+'/'+file_name
+    if not os.path.exists(vocab_dir):
+        os.makedirs(vocab_dir)
+    # Try to read the vocabulary
+    if os.path.isfile(vocab_path):
+        with open(vocab_path, 'rb') as vocab_file:
+            vocab = pickle.load(vocab_file)
+            vocab_file.close()
+            return vocab
+
+    tokens_from_datasets = _get_tokens(input_features,
+                                       tokenizer, 
+                                       datasets,
+                                       preprocessing_fun)
+
+    # Count all the tokens.
+    counter = collections.Counter()
+    for tokens in tokens_from_datasets:
+        counter.update(tokens)
+
+    # Add special tokens to the start of vocab.
+    vocab = collections.OrderedDict()
+    for token in special_tokens:
+        vocab[token] = len(vocab)
+
+    # Add all other tokens to the vocab 
+    for token, freq in sorted(
+            # Sort by frequency (from high to low), and then by token string.
+            # This makes sure high frequency tokens get a low token ID.
+            counter.items(),
+            key=lambda token_freq: (-token_freq[1], token_freq[0])):
+        vocab[token] = len(vocab)
+
+    logging.info('Number of unfiltered tokens: %d', len(counter))
+    logging.info('Vocabulary size: %d', len(vocab))
+
+    # Save the vocabulary to disk
+    with open(vocab_path, 'wb') as vocab_file:
+        pickle.dump(vocab, vocab_file)
+        vocab_file.close()
```

Comment:
If you use a context manager you don't have to manually close the file.

```suggestion
    with open(vocab_path, 'wb') as vocab_file:
        pickle.dump(vocab, vocab_file)
```